migrations/migration.py

```diff
@@ -0,0 +1,178 @@
+            curr_migration = filename[:-3]
+            prev_migration = self.get_previous_migration_id(filename)
+            migration_sequence[curr_migration] = prev_migration
+
+        # Perform DFS from each node
+        for node in migration_sequence:
+            if node not in visited:
+                dfs(node)
+
+        return migration_sequence
+
+    def get_migration_files(self) -> List[str]:
+        """Retrieve the list of migration files."""
+        migration_files = os.listdir(self.migrations_dir)
+        return migration_files
+
+    def get_previous_migration_id(self, filename: str) -> str:
+        """Retrieve the id of the previously ran (downstream) migration script."""
+        prev_migration_id = None
+        with open(os.path.join(self.migrations_dir, filename), "r") as migration_file:
+            for line in migration_file:
+                match = re.match(r"# Previous version: (.+)", line)
+                if match and match.group(1) != 'None':
+                    prev_migration_id = match.group(1)
+                    break
+        return prev_migration_id
+
+    def generate_migration_script(self, migration_name: str):
+        """Generate a new migration script."""
+        current_id = str(self.get_latest_applied_migration_from_fhir())
+        new_id = str(uuid.uuid4())  # Random string as the migration identifier
+        migration_filename = f"{new_id}.py"
+        migration_path = os.path.join(self.migrations_dir, migration_filename)
+        with open(migration_path, "w") as migration_file:
+            migration_file.write(f"# Migration script generated for {migration_name}\n")
+            migration_file.write(f"# Current version: {new_id}\n")
+            migration_file.write(f"# Previous version: {current_id}\n")
+            migration_file.write("\n")
+            migration_file.write("# Add your migration code here\n")
+        return migration_filename
+
+    def run_migrations(self, direction: str):
+        """Run migrations based on the specified direction ("upgrade" or "downgrade")."""
+        if direction not in ["upgrade", "downgrade"]:
+            raise ValueError("Invalid migration direction. Use 'upgrade' or 'downgrade'.")
+
+        current_migration = self.get_latest_applied_migration_from_fhir()
+        applied_migration = None
+
+        if direction == "upgrade":
+            applied_migration = self.get_next_migration(current_migration)
+        elif direction == "downgrade" and current_migration is not None:
+            applied_migration = self.get_previous_migration(current_migration)
+
+        if applied_migration is None:
+            raise ValueError("No valid migration files to run.")
+
+        migration_path = os.path.join(self.migrations_dir, applied_migration)
+        with open(migration_path, "r") as migration_file:
+            migration_code = migration_file.read()
+            try:
+                exec(migration_code)
```

Comment:
`exec` is notoriously dangerous.  we should just use entry points within the file.  please try `__import__(migration_file)` or see the `imp` module which you can use to import the migration file by path.  Then call `upgrade` or `downgrade` directly.